internal/oauth2/lazy_token.go

```diff
@@ -0,0 +1,107 @@
+package oauth2
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+// Tokener is an interface that defines the AccessToken method.
+type Tokener interface {
+	Token(context.Context) (string, error)
+	ForceRefresh(context.Context) (string, error)
+}
+
+type LazyToken struct {
+	// Url represents the URL used for acquiring the token.
+	Url string
+	// ClientId is the client ID used for authentication.
+	ClientId string
+	// ClientSecret is the client secret used for authentication.
+	ClientSecret string
+	// 	AccessToken string holds the currently cached token.
+	AccessToken string
+	// 	RefreshToken string token after refresh.
+	RefreshToken string
+	// Expiration stores the expiration time of the current token.
+	Expiration time.Time
+	// mutex ensures safe concurrent access to the token and expiration fields.
+	mutex sync.Mutex
+}
+
+type tokenResponse struct {
+	AccessToken  string    `json:"access_token"`
+	Expiration   time.Time `json:"expiration"`
```

Comment:
I noticed one thing, this is not how the response actually look like from RH OIDC. It actually looks like this:

```
{
  "access_token": "xxxxxxxx",
  "expires_in": 900,
  "refresh_expires_in": 0,
  "token_type": "Bearer",
  "not-before-policy": 0,
  "scope": "api.console"
}
```

You need to provide a scope I think to [get refresh token](https://auth0.com/docs/secure/tokens/refresh-tokens/get-refresh-tokens), I haven’t tried this myself. Note the fields has also different names (expires_in vs expiration).

Reply:
Ok so I did[ read more about oauth2 token refreshing](https://www.oauth.com/oauth2-servers/access-tokens/access-token-response/) and it appears that refresh token is actually optional, our service does not appear to return it. In that case, we can stick with the initial goal - when access token expires, just request a new one with id/secret. That should not be any problem, this is what IB does for every single request after all.

So in this case, I believe you only need to fix the expiration. The field name is named differently (`expires_in`) and it is not time but amount of seconds from now so you need to convert this or update the code to use seconds instead of time.